text/0000-edition-2021-and-beyond.md

```diff
@@ -0,0 +1,194 @@
+
+## Idiom lint transitions
+
+"Idiom lints" are issued in a lint group named after the edition year, such as `rust_2018_idioms`. They are warn-by-default in the previous edition, and are deny by default in the new edition.
+
+Idiom lints are encouraged but not required to produce "rustfix"-compatible suggestions.
+
+## Keyword reservation policy
+
+One question that comes up around editions is whether to reserve keywords which we think we *might* want but for which we don't have a particular use in mind yet. For the Rust 2018 edition, we opted not to reserve any such keywords, and in this RFC we re-affirm that policy.
+
+The policy is that **new keywords can be introduced in an edition only as part of a design with an accepted RFC**. Note that if there is an accepted RFC for some design that introduces a new keyword, but the design is not yet fully implemented, then the edition might still make that keyword illegal. This way, the way is clear when the time comes to introduce that keyword in the future. As an example, this is what happened with async/await: the async keyword was introduced as part of the 2018 edition, but didn't do anything until later in the release cycle.
+
+The motivation here is that any attempt to figure out a reasonable set of keywords to reserve seems inevitably to turn into "keyword fishing", where we wind up with a long list of potential keywords. This ultimately leads to user confusion and a degraded experience. Given that editions come on a regular basis, it suffices to simply allow the keyword to be reserved in the next edition. If we really want to expose the feature earlier, then a macro or other workaround can be used in the interim (and transitioned automatically as part of the move to the next edition).
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The primary drawbacks of doing editions at all are as follows:
+
+* Coordinating an edition release is a stressor on the organization, as we have to coordinate the transition tooling, documentation, and other changes. This was particularly true in the original formulation of the editions, which put a heavy emphasis on the "feature-driven" nature of the 2018 Edition (i.e., the goal was to release new and exciting features, not to look back on work that had already been completed).
+* Transitioning to a new edition, even if optional, is an ask for our users. Some production users expressed frustration at having to spend the time upgading their crates to the new edition. Even with tooling, the task requires time and effort to coordinate. At this stage in Rust's life, "production use" often implies "commercial use," and time and effort means "money" to them. Asking too much could harm Rust's commercial prospects, with all of the secondary impacts that has on the not-for-profit ecosystem as well. 
```

Comment:
> Some production users expressed frustration at having to spend the time upgading their crates > to the new edition

For me, this is not about upgrading our crates to a new edition, but this frustration is also probably about __upgrading dependencies__ (e.g. random projects where we could raise a PR but it can also be ignored for a very long time) to a new edition (especially if that edition has 0 breaking changes). I think this is a drawback in my view that's not acurately captured in the text.

Reply:
Why would there be a need to upgrade dependencies? Dependencies being on an older edition doesn't prevent crates that depend on those crates from using a newer edition (and vice-versa).

Reply:
When would you *need* a dependency to be upgraded to a new edition? Given that there's edition interoperability, it shouldn't matter, right?

Reply:
Well are there improvements that are edition specific (even with 0 breaking change)? That's not at all clear.

Before thos, 0 breaking change and a compiler upgrade means that all of my dependencies get compiled with the new compiler and get all the benefits of that.

This edition system makes no guarantee that any improvements that are not breaking changes apply to all editions.

Reply:
Put it another way, can things like adding an option for `-ffast-math` or some equivalent option discussed [here](https://github.com/rust-lang/rust/issues/21690) be edition-specific? If the math library library we depend on doesn't change it's edition, how can i enable those (let's say it's a feature available from certain editions forward.

Is this a case that could happen under the edition system? (it would have no breaking change, e.g. it adds a compiler feature flag).

Reply:
I think as someone that's an _in production_ user of Rust, I would rather have stronger guarantees on what can change in an edition and what can't be an edition change, just so I don't have to chase my dependencies to upgrade editions to get the benefits that today i'd get with a compiler upgrade.

Reply:
I have to admit I don't really understand the question. The general goal of the system is that **you don't care, or even know** what edition your dependencies are using. That is basically not relevant to you. What *is* relevant is what version of rustc they require, but that is true whether or not a new edition has been published (i.e., we may stabilize some new feature in rustc verson 1.X, and if they use it, then you must also use a rustc version >= 1.X). There is some talk of adding "minimum rustc version" metadata in Cargo.toml, but it's an orthogonal concern.

(Caveat: there could be some features that are difficult to access without using the new edition. For example, if a crate uses `async fn`, and you wish to interact with it by writing other `async fn`, you both have to be using Rust 2018 to get access to the `async` keyword. These features are quite unusual.)

Reply:
So what we're saying is that editions can't introduce e.g. new optimizations that the compiler emits under the edition flag.

Reply:
It will have no impact on the supported features, but _possibly_ on the supported Rust version.

For example, if you change `edition = "2021"` to `edition = "2024"`, your crate will fail to compile on Rust versions that don't support Rust 2024. However, if you already use features that were introduced _after_ Rust 2024 was released, there's no problem.

Reply:
So it's like a super basic way of saying: _from now on, my code wants this toolchain version that allows 2021 as a rust edition_, even tho 2021 edition has no breaking changes compared to 2018 edition.